---
output:
  pdf_document: default
  html_document: default
params:
    data: "."
    assemble_align: "AssembleAlignLog_table.tab"
    assemble_reference: "AssembleReferenceLog_table.tab"
---

```{r, message=FALSE, echo=FALSE, results="hide"}
# Setup
library(prestor)
library(knitr)
library(captioner)
if (!exists("tables")) { tables <- captioner(prefix="Table") }
if (!exists("figures")) { figures <- captioner(prefix="Figure") }
figures("assemble_prcons", "Count of mate-pairs passing and failing assembly by C-region primer.")
figures("assemble_conscount", "Count of mate-pairs passing and failing assembly by UMI read group size.")
figures("assemble_length", "Distribution of assembled read lengths.")
figures("assemble_error", "Distribution of error rates or identity for assemblies.")
figures("assemble_pvalue", "Distribution of P-value and E-value scores for assemblies.")
```

```{r, echo=FALSE}
assembly_logs <- list()
assembly_logs[[1]] <- loadLogTable(file.path(params$data, params$assemble_align))

ref_eval <- file.exists(file.path(params$data, params$assemble_reference))
if (ref_eval) {
    assembly_logs[[2]] <- loadLogTable(file.path(params$data, params$assemble_reference))
    titles <- c("Align", "Reference")
} else {
    titles <- "Align"
}
```

# Paired-End Assembly

Assembly of paired-end reads (using the AssemblePairs tool) determines the read
overlap in two steps. First, an exhaustive approach is used to identify all possible
overlaps between the two reads with alignment error rates below a user-defined
threshold, where error rates are calculated for each pair of sequence segments using
the same scoring method employed in primer alignment.

## Counts of passing and failing assemblies

```{r, echo=FALSE}
plot_params <- list(titles=titles, style="field", field="PRCONS", sizing="figure")
do.call(plotAssemblePairs, c(assembly_logs, plot_params))
```

`r figures("assemble_prcons")`

```{r, echo=FALSE}
plot_params <- list(titles=titles, style="field", field="CONSCOUNT", sizing="figure")
do.call(plotAssemblePairs, c(assembly_logs, plot_params))
```

`r figures("assemble_conscount")`

## Assembled sequence lengths

```{r, echo=FALSE}
plot_params <- list(titles=titles, style="length", sizing="figure")
do.call(plotAssemblePairs, c(assembly_logs, plot_params))
```

`r figures("assemble_length")`


## Alignment error rates and significance

```{r, echo=FALSE}
plot_params <- list(titles=titles, style="error", sizing="figure")
do.call(plotAssemblePairs, c(assembly_logs, plot_params))
```

`r figures("assemble_error")`

```{r, echo=FALSE}
plot_params <- list(titles=titles, style="pvalue", sizing="figure")
do.call(plotAssemblePairs, c(assembly_logs, plot_params))
```

`r figures("assemble_pvalue")`